Bluetooth: Update audio Object Transfer Client, move alongside Object Transfer Service

asbjornsabo opened the pull request:
An Object Transfer Client (OTC) has been developed and maintained along with the le-audio profiles and services, as the le-audio Media Control Client depends upon OTC to be full-featured.

The purpose of this PR is to make this OTC implementation "official" and available.
The PR moves the OTC implementation from the le-audio code to alongside the corresponding Object Transfer Service.
The OTC code is cleaned up a bit, and is also updated to use the existing OTS definitions and L2CAP library to avoid duplication.

UPDATE: Additional commits to unify client and server more, including merging the two header files.

The OTC is used by the media control client, and is covered by BabbleSim tests for that module.

The initial implementation the OTC was done by Leif Aschehoug <Leif-Alexandre.Aschehoug@nordicsemi.no>
The code has later been updated and maintained by Emil Gydesen emil.gydesen@nordicsemi.no and Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

Note: 
Commits 2 and 3 deal with removing remains of the old OTS implementation that also existed in the le-audio code (and which the starting point for, and was replaced by, the current OTS code).
Commit 4 is the move from the le-audio code to alongside the OTS, in the "bluetooth/services/ots" folder
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/41592

asbjornsabo committed:
Bluetooth: OTC: Add TODO to fix macro

Object Transfer Client: Add TODO to fix the OLCP result conversion
macro

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: OTC: Remove old ots internal header file

This commit is the first in a series of commits to make the OTC
(Object Transfer Client) implementation that exists in the audio
folder "official". That is, the client implementation will be updated
and refactored to be aligned with the OTS (Object Transfer Service)
implementation that exists in subsys/bluetooth/services/ots, and moved
to that folder.

Preparations for making OTC "official":
Move remaining content of old ots_internal header fil into the
otc_internal header file, and remove the ots_internal header file.
(Another ots_internal header file already exists in the ots folder.)

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: OTC - remove old ots header file

Preparations for making OTC "official"

Move remaining content of old ots header fil into the otc header file,
and remove the old ots header file.

(Another ots header file already exists in the ots folder.)

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: OTC: Move OTC files to OTS folder

Move the OTC files to the OTS folders for further refactoring.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: OTC - do not register L2CAP server

Remove L2CAP server registration.  The object transfer client does not
need to register an L2CAP server.

OTS spec: "It is always the responsibility of the Client to open the
Object Transfer Channel, regardless of the direction of transfer, and
is never the responsibility of the Server."

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: OTC: Use ots_l2cap for l2cap transfer

Update the OTC to use the existing ots_l2cap code rather than its own
implementation.
- Add an ots_l2cap struct and register it
- Add ots_l2cap callbacks, reuse most of old l2cap_recv callback in
  the ots_l2cap rx_done callback
- Remove old l2cap code

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: OTC: Remove unused macros and net_buf

Remove some unused code from the OTC.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: OTS: add connect() function to ots_l2cap

Add an l2cap connect() function to OTS.  This is a function for
clients, added to prepare for the upcoming OTC client (re)using the
ots_l2cap module.

As part of this, factor out an internal function to find a free l2cap
context.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: OTC: Clean up internal header file

Remove various unused defines, update comments, update inclusion guard

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: Media control - add missing include

The media control client shell must include the OTC header file.
(It is currently indirectly included via otc_internal.h.)

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: OTS: Move directory listing defines to other header

Move the directory listing size defines to the OTS directory listing
internal header file.

This removes an implicit dependency upon users of
ots_dir_list_internal.h to also include ots_internal.h.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: OTC: Use OTS defines and types instead of separate ones

Update the Object Transfer Client to use OTS defines and types instead
of the ones left over from the previous "le-audio local" object
transfer service:
DIR_LIST, object ID size, OACP and OLCP

Remove the no longer used defines and types from the OTC header file.
Exception: The OLCP sort order defines have not been removed, but
moved to the olcp internal header file, for expected later use.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: OTC: Minor documentation update

Minor updates to documentation in OTC header file.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: OTC: Rename two functions

Rename functions to follow common pattern (module, action, noun).

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented:
@ekohandel  I was not able to add you as a reviewer to the PR.  But since you are one of the users of OTS, I would like you to review this PR anyway.

ekohandel commented:
> @ekohandel I was not able to add you as a reviewer to the PR. But since you are one of the users of OTS, I would like you to review this PR anyway.

@asbjornsabo, I am unable to comment on the `Kconfig.otc` file as it was only renamed but it looks like this file hasn't been generalized yet as it has references to bluetooth audio. Also I think it make sense for `BT_AUDIO` to depend on or select `BT_OTC` instead of other way around?

ekohandel commented on include/bluetooth/services/otc.h:
```diff
@@ -3,24 +3,41 @@
+	uint8_t seconds;
+};
+
+#define BT_OTS_NAME_MAX_SIZE           120
```
Does it make sense to name this `BT_OTC_NAME_MAX_SIZE` to match the use case of it being for the client. Also to differentiate it better from the `BT_OTS_OBJ_MAX_NAME_LEN` Kconfig value.

asbjornsabo replied:
It is used for a maximum limit.  But it is also used in `bt_otc_obj_metadata` which is again used in `bt_otc_instance_t`, which is used publicly, so not straight forward.
May possibly be made non-public if the client and server headers are merged and the instance_t made opaque.

asbjornsabo replied:
Changed to use the Kconfig also used by the server

ekohandel commented on include/bluetooth/services/otc.h:
```diff
@@ -3,24 +3,41 @@
+#define BT_OTS_NAME_MAX_SIZE           120
+
 #define BT_OTC_MAX_WRITE_SIZE               23
 #define BT_OTC_UNKNOWN_ID                   0xFF
```
Looks like majority of these defines are only used in `ots.c` for internal logic. Could they be made "private" by moving them to `ots.c`?

asbjornsabo replied:
Moved the two last ones.

ekohandel commented on include/bluetooth/services/otc.h:
```diff
@@ -106,21 +123,25 @@ struct bt_otc_instance_t {
 	struct bt_otc_obj_metadata cur_object;
```
It seems not ideal that there are two different types for metadata but I can't think of an easy way to combine the two at the moment. Although we could reuse the `struct bt_ots_obj_size` for the current and allocated size metadata fields.

ekohandel commented on include/bluetooth/services/otc.h:
```diff
@@ -106,21 +123,25 @@ struct bt_otc_instance_t {
+typedef int (*bt_otc_dirlisting_cb)(struct bt_otc_obj_metadata *meta);
 
+/** OTC callback structure */
 struct bt_otc_cb {
```
In `struct bt_ots_cb`, the instance parameter comes first and then `struct bt_conn`, should we keep the same convention here for consistency?

asbjornsabo replied:
Done.
Although in many other cases, the connection is the first parameter.  (In which case I should rather have changed the ots.)
@Thalley What is your opinion?

Thalley replied:
Good question. 
On one hand `conn` is the first argument in all other APIs besides OTS, but on the other hand it certainly makes sense to keep the `ots_client` and `ots` APIs similar. 

I'm tempted to suggest that we do `conn, ots` and update the OTS API while we can. 

Another thing to consider, does the OTS (and client) API need the `conn` point at all? For e.g. most LE Audio services we've actually omitted the `conn` pointer in the APIs and simply have an instance pointer. Is there a usecase for having the `bt_conn` pointer at all?

asbjornsabo replied:
It does not seem to be really needed.
Although if we remove it, we may want to add the connection to the instance instead, as for e.g. vcs.
Anyway, changing the API of OTS seems to be to be outside of the scope of getting the OTC in.

asbjornsabo replied:
https://github.com/zephyrproject-rtos/zephyr/issues/42732

ekohandel commented on include/bluetooth/services/otc.h:
```diff
@@ -106,21 +123,25 @@ struct bt_otc_instance_t {
 	struct bt_otc_obj_metadata cur_object;
 };
```
Should we call this structure `bt_otc` to be consistent with `bt_ots`?

asbjornsabo replied:
This is the metadata struct, not the instance struct.
The instance struct is renamed to bt_ots_client.

ekohandel reviewed (changes requested):
@asbjornsabo, it will probably take me a few days to get through this, I work on this as a side project and not part of my day job so I don't have as much time as I would like to put towards it. I will continue to submit my comments in an incremental fashion. Thanks for including me in this review.

asbjornsabo committed:
Bluetooth: OTC - mark as Experimental

Mark the OTC implementation as experimental

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

Thalley commented on subsys/bluetooth/services/ots/otc.c:
```diff
@@ -169,15 +126,91 @@ static int read_attr(struct bt_conn *conn,
-static void print_oacp_response(enum bt_ots_oacp_proc_type req_opcode,
-				enum bt_ots_oacp_res_code result_code)
+/* L2CAP callbacks */
+static void tx_done (struct bt_gatt_ots_l2cap *l2cap_ctx,
```
```suggestion
static void tx_done(struct bt_gatt_ots_l2cap *l2cap_ctx,
```

asbjornsabo replied:
Fixed

Thalley commented on subsys/bluetooth/services/ots/otc.c:
```diff
@@ -169,15 +126,91 @@ static int read_attr(struct bt_conn *conn,
+	BT_ERR("Unexpected call, context: %p, conn: %p", l2cap_ctx, conn);
+}
+
+static ssize_t rx_done (struct bt_gatt_ots_l2cap *l2cap_ctx,
```
```suggestion
static ssize_t rx_done(struct bt_gatt_ots_l2cap *l2cap_ctx,
```

Thalley commented on subsys/bluetooth/services/ots/otc.c:
```diff
@@ -169,15 +126,91 @@ static int read_attr(struct bt_conn *conn,
+{
+	uint32_t offset = cur_inst->rcvd_size;
+	bool is_complete = false;
+	struct bt_otc_obj_metadata *cur_object =
```
```suggestion
	const struct bt_otc_obj_metadata *cur_object =
```

asbjornsabo replied:
Fixed

Thalley commented on subsys/bluetooth/services/ots/otc.c:
```diff
@@ -169,15 +126,91 @@ static int read_attr(struct bt_conn *conn,
+	cur_inst->rcvd_size += buf->len;
+
+	if (cur_inst->rcvd_size >= cur_object->current_size) {
+		is_complete = true;
+	}
+
+	if (cur_inst->rcvd_size > cur_object->current_size) {
+		BT_WARN("Received %u but expected maximum %u",
+			cur_inst->rcvd_size, cur_object->current_size);
+	}
```
Alternatively leave the latter `if` as is, and change `is_complete` to 
```
const bool is_complete = cur_inst->rcvd_size >= cur_object->current_size;
```

asbjornsabo replied:
I can not see that this would do any functional change, so I am afraid I might be missing something.
Is the issue that the two if-s are partly "overlapping", and that you find it cleaner to have the second as a "specialization" of the first?

Thalley commented on subsys/bluetooth/services/ots/otc.c:
```diff
@@ -169,15 +126,91 @@ static int read_attr(struct bt_conn *conn,
+						    is_complete, 0);
+
+	if (is_complete) {
+		uint32_t rcv_size = cur_object->current_size;
```
```suggestion
		const uint32_t rcv_size = cur_object->current_size;
```

asbjornsabo replied:
Fixed

Thalley commented on subsys/bluetooth/services/ots/otc.c:
```diff
@@ -169,15 +126,91 @@ static int read_attr(struct bt_conn *conn,
+
+		BT_DBG("Received the whole object (%u bytes). "
+		       "Disconnecting L2CAP CoC", rcv_size);
+		err = bt_gatt_ots_l2cap_disconnect(l2cap_ctx);
+
+		if (err < 0) {
```
```suggestion
		err = bt_gatt_ots_l2cap_disconnect(l2cap_ctx);
		if (err < 0) {
```

asbjornsabo replied:
Is this some guideline?
Personally, I often prefer a blank line here, I feel it improves readability.
(We have a guideline to have a blank line after a right curly brace - logically it seems to make sense to have one at the start of the block as well.)

Thalley replied:
It is not - But it is something that @jhedberg have often requested from me :) I've added an entry for this in the BT subsystem guidelines discussion: https://github.com/zephyrproject-rtos/zephyr/discussions/41120#discussioncomment-2066740

asbjornsabo replied:
Changed

Thalley commented on subsys/bluetooth/services/ots/otc.c:
```diff
@@ -169,15 +126,91 @@ static int read_attr(struct bt_conn *conn,
+
+		cur_inst = NULL;
+	} else if (cb_ret == BT_OTC_STOP) {
+		uint32_t rcv_size = cur_object->current_size;
```
```suggestion
		const uint32_t rcv_size = cur_object->current_size;
```

asbjornsabo replied:
Fixed

Thalley commented on subsys/bluetooth/services/ots/otc.c:
```diff
@@ -169,15 +126,91 @@ static int read_attr(struct bt_conn *conn,
+
+		BT_DBG("Stopped receiving after%u bytes. "
+		       "Disconnecting L2CAP CoC", rcv_size);
+		err = bt_gatt_ots_l2cap_disconnect(l2cap_ctx);
+
+		if (err < 0) {
```
```suggestion
		err = bt_gatt_ots_l2cap_disconnect(l2cap_ctx);
		if (err < 0) {
```

Thalley commented on subsys/bluetooth/services/ots/otc.c:
```diff
@@ -169,15 +126,91 @@ static int read_attr(struct bt_conn *conn,
+		}
+
+		cur_inst = NULL;
+
```
```suggestion
```

Thalley commented on subsys/bluetooth/services/ots/otc.c:
```diff
@@ -169,15 +126,91 @@ static int read_attr(struct bt_conn *conn,
+	return 0;
+}
+
+static void closed (struct bt_gatt_ots_l2cap *l2cap_ctx,
```
```suggestion
static void closed(struct bt_gatt_ots_l2cap *l2cap_ctx,
```

Thalley commented on subsys/bluetooth/services/ots/otc.c:
```diff
@@ -238,48 +271,48 @@ static void olcp_ind_handler(struct bt_conn *conn,
 
 		switch (req_opcode) {
-		case BT_OTS_OLCP_PROC_FIRST:
+		case BT_GATT_OTS_OLCP_PROC_FIRST:
```
Outside the scope of this PR, but I don't think we really need the `GATT` infix for OTS defines. Could be renamed to `BT_OTS_`.

asbjornsabo replied:
It seems to be quite consequently used in OTS.  But I agree.
Would like to get some "blessing" from e.g. Johan in advance of changing all that.

Thalley replied:
Can be postponed to another PR later. 

At least since OTS isn't in a `GATT` folder, the `BT_GATT` prefix doesn't make much sense to me at least :D

Thalley commented on subsys/bluetooth/services/ots/otc.c:
```diff
@@ -396,26 +429,16 @@ int bt_otc_register(struct bt_otc_instance_t *otc_inst)
-			l2cap_registered = true;
+		err = bt_gatt_ots_l2cap_register(&otc_insts[i].l2cap_ctx);
+		if (err) {
+			BT_WARN("Could not register L2CAP context");
```
```suggestion
			BT_WARN("Could not register L2CAP context: %d", err);
```

Thalley commented on subsys/bluetooth/services/ots/ots_l2cap.c:
```diff
@@ -252,6 +265,42 @@ int bt_gatt_ots_l2cap_unregister(struct bt_gatt_ots_l2cap *l2cap_ctx)
+	int err;
+	struct bt_gatt_ots_l2cap *ctx;
+
+	*l2cap_ctx = NULL;
+
+	if (!conn) {
+		LOG_WRN("Invalid Connection");
+		return -ENOTCONN;
+	}
```
If we NULL-check `conn`, we should NULL check `l2cap_ctx`

Thalley commented on subsys/bluetooth/services/ots/ots_l2cap.c:
```diff
@@ -252,6 +265,42 @@ int bt_gatt_ots_l2cap_unregister(struct bt_gatt_ots_l2cap *l2cap_ctx)
+	}
+
+	l2cap_chan_init(&ctx->ot_chan);
+	memset(&ctx->tx, 0, sizeof(ctx->tx));
```
```suggestion
	(void)memset(&ctx->tx, 0, sizeof(ctx->tx));
```

Thalley commented on subsys/bluetooth/services/ots/CMakeLists.txt:
```diff
@@ -9,3 +9,8 @@ zephyr_sources_ifdef(
 
   zephyr_library_sources_ifdef(CONFIG_BT_OTS_DIR_LIST_OBJ ots_dir_list.c)
+
+zephyr_library_sources_ifdef(
+  CONFIG_BT_OTC
+  otc.c
+  ots_l2cap.c)
```
If `CONFIG_BT_OTC` compiles `ots_l2cap.c`, then I think we should guard the individual functionality based on whether `CONFIG_BT_OTC` is enabled or `CONFIG_BT_OTS` is enabled.

asbjornsabo replied:
Within the file, you mean?

Thalley replied:
Yes, e.g. similar to how `gatt.c` has conditional guards for e.g. the gatt client only functions. 

Not sure if it really does a difference in compile size, as I'd assume that the compiler can optimize it though.

asbjornsabo committed:
Bluetooth: OTS/OTC: Config changes for using l2cap for OTC

Update configuration so that the OTS l2cap file can also be used for
the OTC.
(The plan is to merge the OTS and OTC config files later.)

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: OTS - join Kconfig for OTS server and client

Merge Kconfig file for OTS server and client.
Rename the client config from _OTC to _OTS_CLIENT

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented on subsys/bluetooth/services/Kconfig:
```diff
@@ -15,5 +15,6 @@ rsource "Kconfig.hrs"
 rsource "Kconfig.tps"
 
 rsource "ots/Kconfig"
+rsource "ots/Kconfig.otc"
```
Done

asbjornsabo committed:
Bluetooth: OTS - rename client files

Rename OTS client files from "otc" to "ots_client".

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented on subsys/bluetooth/services/ots/otc.c:
```diff
@@ -20,11 +20,15 @@
-#include "ots.h"
-#include "ots_internal.h"
-#include "otc.h"
+#include <bluetooth/services/ots.h>
+#include <bluetooth/services/otc.h>
+#include "otc_internal.h"
+#include "ots_l2cap_internal.h"
+#include "ots_dir_list_internal.h"
+#include "ots_oacp_internal.h"
+#include "ots_olcp_internal.h"
```
Moved the otc to otc_client.  (May merge header files later.)

asbjornsabo replied:
To ots_client, I mean.

asbjornsabo commented on include/bluetooth/services/otc.h:
```diff
@@ -262,7 +283,7 @@ int bt_otc_select_prev(struct bt_conn *conn,
  *  @return int         0 if success, ERRNO on failure.
  */
-int bt_otc_obj_metadata_read(struct bt_conn *conn,
+int bt_otc_obj_read_metadata(struct bt_conn *conn,
```
Changed to read_object_metadata

asbjornsabo commented on include/bluetooth/services/otc.h:
```diff
@@ -3,24 +3,41 @@
 extern "C" {
 #endif
 
+#define SET_OR_CLEAR_BIT(var, bit_val, set)				\
+	((var) = (set) ? ((var) | bit_val) : ((var) & ~bit_val))
```
Still used, for the METADATA defines

asbjornsabo commented on subsys/bluetooth/services/ots/otc.c:
```diff
@@ -33,61 +37,13 @@
-#define L2CH_WORK(_work) CONTAINER_OF(_work, struct l2ch, recv_work)
-#define L2CAP_CHAN(_chan) _chan->ch.chan
-
 struct dirlisting_record_t {
```
Not sure.
I believe the `ots_dir_list_internal.h` contains what is used by the `ots_dirlist.c`, while this struct is used by the client.
It could make sense to move it there, but should then also the directory listing client code be moved there?

asbjornsabo commented on include/bluetooth/services/otc.h:
```diff
@@ -3,24 +3,41 @@
+
+
+/** @brief Date and Time structure.
+ *  TODO: Move somewhere else - bluetooth.h?
```
Actually, it corresponds to the "Date Time" characteristic from GSS.  But I do not think we have a specific place for those, will move to internal file.

asbjornsabo commented on include/bluetooth/services/otc.h:
```diff
@@ -3,24 +3,41 @@
+/** @brief Date and Time structure.
+ *  TODO: Move somewhere else - bluetooth.h?
+ */
+#define DATE_TIME_FIELD_SIZE 7
```
Moved

asbjornsabo replied:
And moved back - it is an indirect dependency for the public API, broke the build.
(May be able to fix if merging header files.)

asbjornsabo commented on subsys/bluetooth/services/ots/otc.c:
```diff
@@ -169,15 +126,91 @@ static int read_attr(struct bt_conn *conn,
+static ssize_t rx_done (struct bt_gatt_ots_l2cap *l2cap_ctx,
+			 struct bt_conn *conn, struct net_buf *buf)
+{
+	uint32_t offset = cur_inst->rcvd_size;
```
Fixed

asbjornsabo commented on subsys/bluetooth/services/ots/otc.c:
```diff
@@ -387,7 +420,7 @@ static uint8_t read_feature_cb(struct bt_conn *conn, uint8_t err,
 int bt_otc_register(struct bt_otc_instance_t *otc_inst)
 {
-	static bool l2cap_registered;
+	int err;
```
Moved

asbjornsabo committed:
Bluetooth: OTS - minor client cleanups

Small clean-ups to client code: Making variables const, adding a value
to a debug print.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented on subsys/bluetooth/services/ots/otc.c:
```diff
@@ -1070,38 +1093,13 @@ static void write_oacp_cp_cb(struct bt_conn *conn, uint8_t err,
 	int err;
 	uint32_t offset = 0;
 	uint32_t length;
+	struct bt_gatt_ots_l2cap *l2c;
```
Fixed

asbjornsabo commented:
I believe I have fixed most of the comments.
The large remaining thing for this round would be to merge the two header files.

Thalley commented on include/bluetooth/services/ots_client.h:
```diff
@@ -262,11 +276,11 @@ int bt_otc_select_prev(struct bt_conn *conn,
-int bt_otc_obj_metadata_read(struct bt_conn *conn,
-			     struct bt_otc_instance_t *otc_inst,
-			     uint8_t metadata);
+int bt_otc_read_object_metadata(struct bt_conn *conn,
```
Given the rename to `ots_client`, so all occurences of `bt_otc` be named to `bt_ots_client`?

asbjornsabo replied:
Done now - at least in the API

Thalley commented on subsys/bluetooth/audio/mcc.c:
```diff
@@ -20,8 +20,8 @@
-#include "otc.h"
-#include "otc_internal.h"
+#include <bluetooth/services/ots_client.h>
+#include "../services/ots/ots_client_internal.h"
```
What does MCC need from the internal header file?

asbjornsabo replied:
https://github.com/zephyrproject-rtos/zephyr/pull/42445
(There is an issue to fix that one as well.)

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_audio/test_scripts/mcs_mcc.sh:
```diff
@@ -33,7 +33,7 @@ Execute ./bs_${BOARD}_tests_bluetooth_bsim_bt_bsim_test_audio_prj_conf \
 
 Execute ./bs_2G4_phy_v1 -v=${VERBOSITY_LEVEL} -s=${SIMULATION_ID} \
-  -D=2 -sim_length=20e6 $@
+  -D=2 -sim_length=30e6 $@
```
Have the L2CAP errors when running mcs_mcc.sh and `media_controller.sh` been fixed as part of this refactor, or are they still present?

asbjornsabo replied:
Will be fixed in https://github.com/zephyrproject-rtos/zephyr/pull/42445

Thalley reviewed (changes requested):
Great job :)

asbjornsabo committed:
Bluetooth: OTS: Move defines to internal file

Move defines from public API to implementation (only used there).

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

ekohandel commented on include/bluetooth/services/ots_client.h:
```diff
@@ -1,28 +1,51 @@
 /* @file
- * @brief Object Transfer client header
+ * @brief Object Transfer Service client header
```
```suggestion
 * @brief Object Transfer Service Client header
```

asbjornsabo replied:
@Thalley Do we have a tradition on this?  I capitalized the OTS, as that is the name of the service.

Thalley replied:
We don't, and `client` is more prevailent than `Client`. The only place where `Client` is used is actually for the previously named `OTC`, so I understand why this comment was made :)

asbjornsabo replied:
Well, now the file is removed anyway.

ekohandel commented on subsys/bluetooth/services/ots/ots_client_internal.h:
```diff
@@ -0,0 +1,44 @@
+
+
+#define OTS_FEATURE_LEN        (uint32_t)(2 * sizeof(uint32_t))
+#define OTS_SIZE_LEN           (uint32_t)(2 * sizeof(uint32_t))
```
```suggestion
#define OTS_SIZE_LEN           (uint32_t)(sizeof(struct bt_ots_obj_size))
```

ekohandel commented on subsys/bluetooth/services/ots/ots_l2cap.c:
```diff
@@ -252,6 +265,42 @@ int bt_gatt_ots_l2cap_unregister(struct bt_gatt_ots_l2cap *l2cap_ctx)
+	memset(&ctx->tx, 0, sizeof(ctx->tx));
+
+	LOG_DBG("Connecting L2CAP CoC");
+	err = bt_l2cap_chan_connect(conn, &ctx->ot_chan.chan, BT_GATT_OTS_L2CAP_PSM);
+
+	if (err < 0) {
```
I think we should keep the `err < 0` condition. I assume this was a typo @Thalley  and you meant to suggest removing the extra space?

Thalley replied:
I meant to remove that, just to be consistent with the other `if (err)` checks, but isn't required.

asbjornsabo replied:
Removed blank line.
Removed " < 0", for consistency with the rest of the file.

ekohandel commented on subsys/bluetooth/services/ots/ots_l2cap.c:
```diff
@@ -252,6 +272,47 @@ int bt_gatt_ots_l2cap_unregister(struct bt_gatt_ots_l2cap *l2cap_ctx)
+	err = bt_l2cap_chan_connect(conn, &ctx->ot_chan.chan, BT_GATT_OTS_L2CAP_PSM);
+
+	if (err < 0) {
+		LOG_WRN("Unable to connect to psm %u (err %u)", BT_GATT_OTS_L2CAP_PSM, err);
```
```suggestion
		LOG_WRN("Unable to connect to psm %u (err %d)", BT_GATT_OTS_L2CAP_PSM, err);
```

ekohandel commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -1,7 +1,7 @@
  *  @brief Bluetooth Object Transfer Client
  *
- * Copyright (c) 2020 - 2021 Nordic Semiconductor ASA
+ * Copyright (c) 2020 - 2022 Nordic Semiconductor ASA
```
This line appears at the top of the file but I cannot comment on it as it is too far away from changes:
```c
/* TODO: Temporarily here - clean up, and move alongside the Object Transfer Service */
```
I believe it can be removed now?

ekohandel commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -169,15 +129,89 @@ static int read_attr(struct bt_conn *conn,
+	}
+
+
+	cb_ret = cur_inst->otc_inst->cb->content_cb(conn, offset,
```
Probably doesn't need to be done as part of this PR but would be nice to return the number of bytes remaining in this read to stay consistent with the server implementation.

asbjornsabo replied:
https://github.com/zephyrproject-rtos/zephyr/projects/38

ekohandel commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -169,15 +129,89 @@ static int read_attr(struct bt_conn *conn,
+
+		BT_DBG("Received the whole object (%u bytes). "
+		       "Disconnecting L2CAP CoC", rcv_size);
+		err = bt_gatt_ots_l2cap_disconnect(l2cap_ctx);
```
Is it strictly necessary to closet the L2CAP connection? It would be feasible to use the same connection to read more objects.

asbjornsabo replied:
This is "inherited" code, so I am frankly not sure.
I guess it would be possible to keep it open.  But we do not know neither whether or when it will be needed again, so just as neat to close it.
Besides, the server side closes the other end of the connection anyway.

ekohandel replied:
I think the server only closes the connection if there is an error, otherwise it leaves it up to the client to manage the connection. But good point about not knowing when to need it again.

Thalley replied:
We could close it using a timer, but that's outside the scope of this PR. Consider creating an issue for it.

ekohandel commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -387,35 +421,24 @@ static uint8_t read_feature_cb(struct bt_conn *conn, uint8_t err,
+
 	return -ENOMEM;
 }
 
```
How does one use the `bt_otc_unregister`? The instance index is not returned from the `bt_otc_register`, maybe this should also take a `struct bt_otc_instance_t *otc_inst` instead of `index` as a parameter?

Also should the l2cap context also be unregistered?

Though, this function is not used by anyone so maybe it can be resolved in a follow up PR.

asbjornsabo replied:
https://github.com/zephyrproject-rtos/zephyr/issues/42649

ekohandel commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -1115,16 +1113,21 @@ static int oacp_read(struct bt_conn *conn,
 	/* TODO: How do we ensure that the L2CAP is connected in time for the
```
We might need to add a small state machine here? I had a related question about not closing the L2CAP connection immediately after a read or write is done.

asbjornsabo replied:
https://github.com/zephyrproject-rtos/zephyr/issues/42647

ekohandel reviewed (changes requested):
@asbjornsabo, I am not really in the loop of the plans for this client but based on some comments it looks like we plan to do a few more passes to clean things up?

asbjornsabo commented:
> @asbjornsabo, I am not really in the loop of the plans for this client but based on some comments it looks like we plan to do a few more passes to clean things up?

It needs a bit more work, yes.  I am not sure at which point we will be satisfied for now, though.
The one thing I will try to do is to merge the header files into one, that should help with some of the above comments.

kapi-no commented on include/bluetooth/services/ots_client.h:
```diff
@@ -106,21 +129,25 @@ struct bt_otc_instance_t {
+typedef int (*bt_otc_dirlisting_cb)(struct bt_otc_obj_metadata *meta);
 
+/** OTC callback structure */
 struct bt_otc_cb {
```
I would also rename callbacks in this structure.

content_cb -> obj_content_received / obj_content_recv
metadata_cb -> obj_metadata_received / obj_metadata_recv

kapi-no commented on include/bluetooth/services/ots_client.h:
```diff
@@ -1,28 +1,51 @@
+#define ZEPHYR_INCLUDE_BLUETOOTH_SERVICES_OTS_CLIENT_H_
+
+/**
+ * @brief Object Transfer Service Client (OTC)
```
Could we add OTS discovery logic to this module? What do you think about this idea?

Then we can get rid of the bt_otc_indicate_handler API and use some other mechanism to notify API users that the OTS  client is ready (ots_client_ready_cb?).

asbjornsabo replied:
Something should be done with the discovery, yes.  But I would like to think about that.
The OTS is used as a secondary service, which complicates a bit.  But @Emil-Gydesen-Bose has as solution for included services in e.g. VCS, I believe.

asbjornsabo replied:
Created https://github.com/zephyrproject-rtos/zephyr/issues/42640

kapi-no reviewed (changes requested):
I focused on the client API in my first review round

asbjornsabo committed:
Bluetooth: OTS: Unify object name size for client and server

Use Kconfig value for object name size also for client.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented on include/bluetooth/services/ots_client.h:
```diff
@@ -1,28 +1,51 @@
 
-#define BT_OTC_MAX_WRITE_SIZE               23
-#define BT_OTC_UNKNOWN_ID                   0xFF
+#define SET_OR_CLEAR_BIT(var, bit_val, set)				\
```
I know - already removing

asbjornsabo committed:
Bluetooth: OTC - Move metadata bitfields to common files

Move the metadata request bitfields to the common header files.
This is part of merging the ots_client and ots header files.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: OTS: Merge client and server metadata types

Merge the metadata types for client and server.
This is part of merging the ots_client and ots header files.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

Thalley commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -20,45 +18,25 @@
 #include <bluetooth/l2cap.h>
 
-#include "../host/conn_internal.h"  /* To avoid build errors on use of struct bt_conn" */
+#include "../../host/conn_internal.h"  /* To avoid build errors on use of struct bt_conn" */
```
This is in order to log `conn` pointer values? If so, then perhaps it's just cleaner to `void *` cast them instead.

asbjornsabo replied:
That works

Thalley commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -20,45 +18,25 @@
-	struct bt_l2cap_le_chan ch;
-};
-
-#define L2CH_CHAN(_chan) CONTAINER_OF(_chan, struct l2ch, ch.chan)
-#define L2CH_WORK(_work) CONTAINER_OF(_work, struct l2ch, recv_work)
-#define L2CAP_CHAN(_chan) _chan->ch.chan
+#define BT_OTC_MAX_WRITE_SIZE    23
```
Inconsistent naming scheme

Thalley replied:
Whoa, the marked code exploded... What I meant was that `OTS_CLIENT_INST_COUNT` and `BT_OTC_MAX_WRITE_SIZE` does not follow the same naming scheme

asbjornsabo replied:
Now they do :-)

asbjornsabo committed:
Bluetooth: OTS - Merge server and client header files

- Move content of client header file to common header file.
- Rename content from "otc" to "ots_client".
- Remove empty client header file.
- Update users.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: OTS - Rename client callbacks

Rename client callbacks to be more consistent with common code style.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented on include/bluetooth/services/ots_client.h:
```diff
@@ -106,21 +129,25 @@ struct bt_otc_instance_t {
 	struct bt_otc_obj_metadata cur_object;
 };
```
That is something we should do, I think.
I am tempted to postpone for now - there is already 25 commits in this PR.

asbjornsabo commented on include/bluetooth/services/ots_client.h:
```diff
@@ -106,21 +129,25 @@ struct bt_otc_instance_t {
  *                stop the decoding.
  */
-typedef int (*bt_otc_dirlisting_cb)(struct bt_otc_obj_metadata *object);
+typedef int (*bt_otc_dirlisting_cb)(struct bt_otc_obj_metadata *meta);
```
I believe I have fixed this everywhere in the API now.

asbjornsabo commented on include/bluetooth/services/ots_client.h:
```diff
@@ -1,28 +1,51 @@
+/** @brief Date and Time structure.
+ *  TODO: Move somewhere else - bluetooth.h?
+ */
+#define DATE_TIME_FIELD_SIZE 7
+struct bt_date_time {
```
Renamed

asbjornsabo commented:
[Thalley](https://github.com/Thalley) [5 hours ago](https://github.com/zephyrproject-rtos/zephyr/pull/42086#discussion_r802487850)

We could close it using a timer, but that's outside the scope of this PR. Consider creating an issue for it.

@Thalley Why do you want to close it with a timer?

asbjornsabo committed:
Bluetooth: OTS - rename two defines

Rename two defines for internal consistency in file.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: OTS - do not include internal connection header

Remove #include of conn_internal.h - cast conn pointers to void for
logging instead.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: OTS - Client APIs to have instance as first parameter

Move the instance as the first parameter of the client function calls
and callbacks, for consistency with the server implementation.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented:
@ekohandel Ready for review again (could not re-ask you the normal way).

Thalley commented:
> 



> [Thalley](https://github.com/Thalley) [5 hours ago](https://github.com/zephyrproject-rtos/zephyr/pull/42086#discussion_r802487850)
> 
> We could close it using a timer, but that's outside the scope of this PR. Consider creating an issue for it.
> 
> @Thalley Why do you want to close it with a timer?

Not sure how this out of the original thread for this comment, but if the case for keeping it open was in case that we needed it again, and we won't know when, then perhaps a e.g. 10 second timer to keep it open (in case the client wants to read more objects), and close it after that, as to not keep it alive for too long unnecessarily. This would be similar to the HTTP keep-alive: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Keep-Alive

Thalley commented on include/bluetooth/audio/mcc.h:
```diff
@@ -923,9 +923,9 @@ int bt_mcc_otc_read_current_group_object(struct bt_conn *conn);
 
 #if defined(CONFIG_BT_MCC_SHELL)
-struct bt_otc_instance_t *bt_mcc_otc_inst(void);
+struct bt_ots_client *bt_mcc_otc_inst(void);
```
```suggestion
struct bt_ots_client *bt_mcc_ots_client_inst(void);
```

Thalley replied:
Can be done in a separate PR, as it is a bit unrelated to this PR

asbjornsabo replied:
Will most likely be fixed when the server and client instances are unified.

Thalley replied:
> Will most likely be fixed when the server and client instances are unified.

Do we have an issue for that?

asbjornsabo replied:
https://github.com/zephyrproject-rtos/zephyr/issues/42642

Thalley commented on include/bluetooth/services/ots.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020 Nordic Semiconductor ASA
+ * Copyright (c) 2020 - 2022 Nordic Semiconductor ASA
```
I think this is the typical way of doing it

```suggestion
 * Copyright (c) 2020-2022 Nordic Semiconductor ASA
```

Thalley commented on include/bluetooth/services/ots.h:
```diff
@@ -468,6 +470,76 @@ struct bt_ots_feat {
 } __packed;
 
+/** @brief Object metadata request bit field values */
+enum {
```
Consider adding a name, and use `enum <name>` in the argument where it's used (not necessary, as we have plenty of APIs that don't do this).

asbjornsabo replied:
Prefer not to - as this is a bitfield, combinations of the values are also valid.
Perhaps it would have been better to find another way of documenting these - it is an enum just to be able to attach a Doxygen comment to it.

Thalley replied:
Hmm, I see. 

We have used (unnamed) enums for bitfields before  (e.g. in bluetooth.h), so that's not really an issue, but you just don't prefer to refer to this enum in the function arguments? I don't have a strong opinion here, but personally I think it's fine to use a named enum for bitfields.

asbjornsabo replied:
IMO, using an enum as the type for the argument communicates "use one of these values".  While in fact a lot of other values (combinations of the values) are also valid.
They values are referred to in the documentation for the function, though.

Alternatively, I could add a Doxygen group (I think it was you suggested).

Thalley commented on include/bluetooth/services/ots.h:
```diff
@@ -468,6 +470,76 @@ struct bt_ots_feat {
+	uint8_t minutes;
+	uint8_t seconds;
+};
+#define DATE_TIME_FIELD_SIZE 7
```
Could this be moved to an internal header? 

```suggestion
#define BT_OTS_DATE_TIME_FIELD_SIZE 7
```

Thalley commented on include/bluetooth/services/ots.h:
```diff
@@ -468,6 +470,76 @@ struct bt_ots_feat {
+ */
+struct bt_ots_obj_metadata {
+
+#if defined(CONFIG_BT_OTS)
+	/** @brief Object Name */
+	char                           *name;
+#endif /* CONFIG_BT_OTS */
+
+#if defined(CONFIG_BT_OTS_CLIENT)
+	/* TODO: Unify client/server name */
+	/** @brief Object name (client) */
+	char                           name_c[CONFIG_BT_OTS_OBJ_MAX_NAME_LEN + 1];
+#endif /* CONFIG_BT_OTS_CLIENT */
```
Could this be a union?

asbjornsabo replied:
It could, but I prefer not to.  It would be quite confusing, I think, to have two identically named variables where one is a pointer and the other a character array.  (I had some errors due to that while refactoring this.)
Propose to leave as is for now, plan is to be able to have them equal.

Thalley commented on subsys/bluetooth/services/ots/Kconfig:
```diff
@@ -1,6 +1,6 @@
 # Bluetooth Object Transfer service
 
-# Copyright (c) 2020 Nordic Semiconductor ASA
+# Copyright (c) 2020 - 2022 Nordic Semiconductor ASA
```
```suggestion
# Copyright (c) 2020-2022 Nordic Semiconductor ASA
```

Thalley commented on subsys/bluetooth/services/ots/Kconfig:
```diff
@@ -69,6 +59,46 @@ config BT_OTS_OLCP_GO_TO_SUPPORT
+	select BT_GATT_CLIENT
+	select BT_GATT_AUTO_DISCOVER_CCC
+	select EXPERIMENTAL
+	default n
```
```suggestion
```

Thalley commented on subsys/bluetooth/services/ots/Kconfig:
```diff
@@ -69,6 +59,46 @@ config BT_OTS_OLCP_GO_TO_SUPPORT
+config BT_DEBUG_OTS_CLIENT
+	bool "Object Transfer Service Client debug"
+	help
+	  Use this option to enable Object Transfer Client debug logs for the
+	  Bluetooth Audio functionality.
```
```suggestion
	  Use this option to enable Object Transfer Client debug logs.
```

Thalley commented on subsys/bluetooth/services/ots/Kconfig:
```diff
@@ -69,6 +59,46 @@ config BT_OTS_OLCP_GO_TO_SUPPORT
+
+if BT_OTS_CLIENT
+
+# TODO: Unify logging for OTS server and client
```
This may need to be more specific, as this could also mean that you'd only want a single Kconfig to enable both.

asbjornsabo replied:
Fixed

Thalley commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -1,13 +1,11 @@
  *  @brief Bluetooth Object Transfer Client
  *
- * Copyright (c) 2020 - 2021 Nordic Semiconductor ASA
+ * Copyright (c) 2020 - 2022 Nordic Semiconductor ASA
```
```suggestion
 * Copyright (c) 2020-2022 Nordic Semiconductor ASA
```

asbjornsabo replied:
Fixed

Thalley commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -169,15 +125,88 @@ static int read_attr(struct bt_conn *conn,
+		BT_WARN("Received %u but expected maximum %u",
+			cur_inst->rcvd_size, cur_object->size.cur);
+	}
+
+
```
```suggestion

```

Thalley commented on subsys/bluetooth/services/ots/ots_internal.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020 Nordic Semiconductor ASA
+ * Copyright (c) 2020 - 2022 Nordic Semiconductor ASA
```
```suggestion
 * Copyright (c) 2020-2022 Nordic Semiconductor ASA
```

Thalley reviewed (dismissed):
At this point it's mostly nit-pick from my side, and I think this is fine to go in, assuming that we might still update the API and/or fix any bugs that may arrive. We probably also want to add standalone OTS+client BSIM tests, so that we get a better test coverage and understanding of the API. 

We should also make a complementary client sample for the OTS sample.

tejlmand commented on subsys/bluetooth/services/ots/Kconfig:
```diff
@@ -69,6 +59,45 @@ config BT_OTS_OLCP_GO_TO_SUPPORT
+config BT_OTS_CLIENT
+	bool "Object Transfer Service Client [Experimental]"
+	select BT_L2CAP_DYNAMIC_CHANNEL
+	select BT_GATT_CLIENT
```
> but I've seen equally many requests to avoid depends on as that hides the settings from the menuconfig GUI.

If settings are forming a hierachy, then `depends on` is usually the right choice.

Just think of it as:
```
config FOO
    bool

if FOO

config FOO_SUBFEATURE
    bool

endif
```

that is identical to.
```
config FOO
    bool

config FOO_SUBFEATURE
    bool
    depends on FOO
```

because `BT_GATT_CLIENT` and `BT_OTS_CLIENT` are on parallel paths in the tree, you don't see this as clear.

The risk of mixing, just because some don't like `that hides the settings from the menuconfig GUI.` is risky because it may easily lead to circular dependencies. 
As well as the `select` bears a higher risk of the stuck-symbol syndrome.

Thalley replied:
@tejlmand Not related to this PR, but for e.g. Bluetooth LE Audio, we are probably looking at something like 10-20 dependencies just for that feature. In some ways doing something like `CONFIG_BT_AUDIO_HEARING_AID=y` instead of enabling those 10-20 dependencies is perhaps also more user friendly (although by hiding away a lot of what has been enabled). 

In such as case, would you be in favor of `depends on` or `select`? 

I'm OK with using `depends on` where possible instead of `select`, as that is more intuitive for me (but I also never use the GUI).

asbjornsabo replied:
@tejlmand What was the conclusion here?  Is this something I should change?
I guess that with "depends" the person doing the configuration would need to _know_ all the dependencies, in order to be able to turn on the feature?  That seems unlikely to be the case.

tejlmand replied:
>  but for e.g. Bluetooth LE Audio, we are probably looking at something like 10-20 dependencies just for that feature. In some ways doing something like CONFIG_BT_AUDIO_HEARING_AID=y instead of enabling those 10-20 dependencies is perhaps also more user friendly (although by hiding away a lot of what has been enabled).

I'm not against `selects` in general, they can make a lot of sense.
What I was asking about, was in cases where something looks like a configuration tree, like:
```
--- BT
      |--- GATT_CLIENT
            |--- BT_OTS_CLIENT
```

but because the tree in Kconfig is actually constructed as this:
```
--- BT
      |--- GATT_CLIENT
      |--- BT_OTS_CLIENT
```

it becomes less visible.
It could be i'm mistaking, and you really want gatt client and ots client on same level, but to me, in this particular case it more looks as OTS client is depending on GATT client.

For the audio, I don't believe those 10-20 dependencies are forming something that could look like a heriachy.
More settings on a common level, correct ?
Hope this answers your question @Thalley 

> @tejlmand What was the conclusion here?

I don't think there is a conclusion yet.
This was more an observation, and i'm trying to understand the reason.
Also, if this is a general way things have been done in bluetooth, then I don't believe a single PR should deviate from that.
But then it could be somethingto consider if structure is correct.

>  Is this something I should change?

I don't think so atm.
This is also why I just posted this as a comment, to try to understand this part better.

asbjornsabo commented:
> 

Added to https://github.com/zephyrproject-rtos/zephyr/issues/42647

asbjornsabo commented:
> > 
> 
> > [Thalley](https://github.com/Thalley) [5 hours ago](https://github.com/zephyrproject-rtos/zephyr/pull/42086#discussion_r802487850)
> > We could close it using a timer, but that's outside the scope of this PR. Consider creating an issue for it.
> > @Thalley Why do you want to close it with a timer?
> 
> Not sure how this out of the original thread for this comment, but if the case for keeping it open was in case that we needed it again, and we won't know when, then perhaps a e.g. 10 second timer to keep it open (in case the client wants to read more objects), and close it after that, as to not keep it alive for too long unnecessarily. This would be similar to the HTTP keep-alive: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Keep-Alive

Added to https://github.com/zephyrproject-rtos/zephyr/issues/42647

ekohandel commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -20,72 +18,29 @@
-#define L2CH_CHAN(_chan) CONTAINER_OF(_chan, struct l2ch, ch.chan)
-#define L2CH_WORK(_work) CONTAINER_OF(_work, struct l2ch, recv_work)
-#define L2CAP_CHAN(_chan) _chan->ch.chan
+#define OTS_CLIENT_MAX_WRITE_SIZE    23
```
I have to admit my knowledge of this area is limited but this seems to be a define that should come from the MTU or maybe a GATT Write size or something? Maybe a follow up clean up item but I see many defines for this magical 23: `BT_ATT_DEFAULT_LE_MTU`, `L2CAP_LE_MIN_MTU`, `BT_RFCOMM_SIG_MIN_MTU` etc.

Thalley replied:
`OTS_CLIENT_MAX_WRITE_SIZE` covers the maximum length of any of the client write operations. This is a static value based on the based, and won't be affected by the MTU, and is just used to define a static write buffer

ekohandel commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -20,72 +18,29 @@
-#define L2CH_WORK(_work) CONTAINER_OF(_work, struct l2ch, recv_work)
-#define L2CAP_CHAN(_chan) _chan->ch.chan
+#define OTS_CLIENT_MAX_WRITE_SIZE    23
+#define OTS_CLIENT_UNKNOWN_ID      0xFF
```
Given that the object ID is only 48bits and we are using 64bits to represent it should we use 0x1000000000000 or maybe even just -1 for this? It is guaranteed to be outside of valid OTS ID ranges. 0xFF is technically reserved for future use.

Thalley replied:
We can't use `-1` as that is a signed value, and since we need to support unsigned 64 bit values, we can't use signed values. 

0xFF is reserved for future use, but using an RFU value is the only way we can store an unknown ID. The alternative here would be to store a boolean value in the struct to determine if it is unknown/uninitialized. 

If we keep going with 0xFF, then perhaps it should be `0x00000000000000FFULL`

asbjornsabo replied:
Changed in additional commit

ekohandel commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -238,48 +266,48 @@ static void olcp_ind_handler(struct bt_conn *conn,
-		case BT_OTS_OLCP_PROC_GOTO:
+		case BT_GATT_OTS_OLCP_PROC_GOTO:
 			BT_DBG("Goto");
 			on_object_selected(conn, result_code, otc_inst);
```
I believe with the GOTO procedure we know the object ID and we shouldn't reset it to an unknown ID. But again this can probably be fixed in a follow up PR.

asbjornsabo replied:
https://github.com/zephyrproject-rtos/zephyr/issues/42782

asbjornsabo committed:
Bluetooth: OTS: Use out-of-range value for unknown object ID

Use a value for "Unknown object ID" that is outside of the range of
valid object IDs.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented:
What needs to be done now to get this through?
@Thalley Your approval seems to have been dismissed - not sure why
@ekohandel Please resolve the comments where you are satisfied with the changes, so that I can more clearly see what is left.
@tejlmand Please conclude on the Kconfig commnt.

Thalley commented:
> What needs to be done now to get this through? @Thalley Your approval seems to have been dismissed - not sure why 

There has been a recent Github change from the TSC, so that if a new commit is added to a PR, all current reviews are invalidated to trigger re-reviews :)

kapi-no commented on include/bluetooth/services/ots.h:
```diff
@@ -708,6 +780,230 @@ int bt_ots_init(struct bt_ots *ots, struct bt_ots_init *ots_init);
+#define BT_OTS_STOP                       0
+#define BT_OTS_CONTINUE                   1
+
+/* TODO: Merge server and client instance as opaque type */
```
Could we move the Client specific API bt_ots_client to a different file for better readability? The OTS client header can include the OTS Server header in which we can host common structures, enums etc.

This header file has over 1000 lines of code. In my opinion, having a dedicated file for the Client module would reduce the users confusion

asbjornsabo replied:
Well, I just spent quite a bit of work to merge the two header files, as that was requested in the PR and that is how it has been done for many of the audio services. Other files, like e.g. gatt.h are the same, I believe.
So I am reluctant to split the header file again - that would have to be as a result of some overall policy with backing from many stakeholders.

kapi-no commented on include/bluetooth/services/ots.h:
```diff
@@ -708,6 +780,230 @@ int bt_ots_init(struct bt_ots *ots, struct bt_ots_init *ots_init);
+int bt_ots_client_read_object_data(struct bt_conn *conn,
+				   struct bt_ots_client *otc_inst);
+
+/** @brief Directory listing object metadata callback
+ *
+ * If a directory listing is decoded using bt_ots_client_decode_dirlisting(),
+ * this callback will be called for each object in the directory listing.
+ *
+ *  @param meta The metadata of the decoded object
+ *
+ *  @return int   BT_OTS_STOP or BT_OTS_CONTINUE. BT_OTS_STOP can be used to
+ *                stop the decoding.
+ */
+typedef int (*bt_ots_client_dirlisting_cb)(struct bt_ots_obj_metadata *meta);
+
+/** @brief Decode Directory Listing object into object metadata.
+ *
+ *  If the Directory Listing object contains multiple objects, then the
+ *  callback will be called for each of them.
+ *
+ *  @param data        The data received for the directory listing object.
+ *  @param length      Length of the data.
+ *  @param cb          The callback that will be called for each object.
+ */
+int bt_ots_client_decode_dirlisting(uint8_t *data, uint16_t length,
+				    bt_ots_client_dirlisting_cb cb);
```
Can we remove these APIs and make decoding automatic inside the implementation? The metadata of each object can be reported with _obj_metadata_read_ callback.

asbjornsabo replied:
That is a possibility, although there is a difference between reading the directory object and the metadata of other object.
Propose to postpone for now: https://github.com/zephyrproject-rtos/zephyr/issues/42819

kapi-no commented on subsys/bluetooth/services/ots/ots_client.c:
```diff
@@ -1,13 +1,11 @@
  *  @brief Bluetooth Object Transfer Client
  *
- * Copyright (c) 2020 - 2021 Nordic Semiconductor ASA
+ * Copyright (c) 2020-2022 Nordic Semiconductor ASA
```
This single implementation file could be divided into smaller implementation units (similar to the OTS server implementation). Potential candidates:

- Module for object navigation ots_olcp_client.c
- Module for object manipulation ots_oacp_client.c

asbjornsabo replied:
That would make for an implementation more similar to the OTS, and could be a good idea.
Propose to postpone for now: https://github.com/zephyrproject-rtos/zephyr/issues/42820

asbjornsabo commented on include/bluetooth/services/ots.h:
```diff
@@ -708,6 +780,230 @@ int bt_ots_init(struct bt_ots *ots, struct bt_ots_init *ots_init);
+ *
+ *  @return int         0 if success, ERRNO on failure.
+ */
+int bt_ots_client_select_id(struct bt_conn *conn,
```
Fixed.
This was already done for the client callbacks, but the functions were forgotten.
Put these changes into that commit.

asbjornsabo commented on subsys/bluetooth/services/ots/Kconfig:
```diff
@@ -69,6 +59,45 @@ config BT_OTS_OLCP_GO_TO_SUPPORT
+
+#### Object Transfer Service Client ################################
+
+config BT_OTS_CLIENT
```
I have not seen anything forbidding a device from being both server and client at the same time, so I do not see why we should.
(We do not restrict this for the audio services either, I believe.)

asbjornsabo committed:
Bluetooth: OTS - rename client data receive callbacks

Rename the data and metadata callbacks to _read(), for consistency
with the corresponding function calls.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented on include/bluetooth/services/ots.h:
```diff
@@ -708,6 +780,230 @@ int bt_ots_init(struct bt_ots *ots, struct bt_ots_init *ots_init);
+	 *  @return int          BT_OTS_STOP or BT_OTS_CONTINUE. BT_OTS_STOP can
+	 *                       be used to stop reading.
+	 */
+	int (*obj_content_recv)(struct bt_ots_client *ots_inst,
```
Agree - fixed

asbjornsabo commented:
Rebased on main with no changes at all.

asbjornsabo commented:
Rebase with no changes again.
The previous rebase was intended to get in the fix for the failing mesh test, but for some reason that did not happen, so the test was still failing.
